Add search functionality for biocontext and biotools APIs

slobentanzer opened the pull request:
- Fix default search to respect registry's impact-first ranking
- Remove alphabetical re-sorting that was overriding impact scores
- Ensure high-impact tools (like SCANPY with 4146 citations) appear at top
- Maintain proper ranking for --sort impact and --type biotools searches
- Combined search now preserves registry ranking order instead of overriding it

peymanvahidi committed:
Adding flag for a detailed status

- Users can understand the reason of their validation fails
- Added some new tests for the new feature

peymanvahidi committed:
Adding support for annotation using relative csv file

slobentanzer committed:
add docs page to structure the approach

slobentanzer committed:
feat: implement registry infrastructure for BioContext integration

- Add RegistryManager class for handling registry operations with caching
- Add BioContextRegistry class for BioContext-specific registry handling
- Implement efficient caching with configurable duration (1 hour default)
- Add comprehensive unit tests with mocked network requests
- Refactor to proper module structure (manager.py, biocontext.py)
- Support case-insensitive search across name, description, and keywords
- Add error handling for network issues and JSON parsing
- Test with real BioContext registry (found 3 PubMed servers)

This provides the foundation for biotope search functionality and future
registry integrations. All tests passing with 11/11 success rate.

slobentanzer committed:
feat: implement biotope search command for MCP registry integration

- Add search command with rich table output for MCP servers
- Integrate with BioContext registry using project configuration
- Add registry settings to init command project structure
- Update BioContextRegistry to accept configurable registry URLs
- Add comprehensive test suite with 9 test cases (all passing)
- Create API documentation with examples and usage instructions
- Update main documentation to include registry integration workflow

The search command allows users to find MCP servers from the BioContext
registry with beautiful formatting and proper error handling. Registry
configuration is automatically set up during project initialization.

slobentanzer committed:
fix tests

slobentanzer committed:
add search to toc

slobentanzer committed:
feat: add MCP registry status to biotope status command

- Add MCP Registry section showing configured registries and accessibility
- Display registry URLs and status (✅ Accessible / ❌ Unreachable)
- Provide contextual suggestions based on registry status
- Add comprehensive unit tests with 8 test cases covering all scenarios
- Test network accessibility, config errors, timeouts, and edge cases
- Only show MCP section when registries are configured

The status command now provides visibility into MCP registry configuration
and connectivity, helping users understand their registry setup and get
guidance on next steps like running 'biotope search <query>'.

slobentanzer committed:
update lock file

slobentanzer committed:
add .biotope/cache to new project's gitignore

slobentanzer committed:
feat: Add GitHub star count ranking to biotope search

- Add star count column to search results table
- Implement GitHub API integration for fetching repository stars
- Add sorting options: relevance (default), stars (descending), name (alphabetical)
- Add --sort option to search command with choices: relevance, stars, name
- Implement graceful error handling for network timeouts and API failures
- Add comprehensive test coverage for star fetching and sorting logic
- Update API documentation with new sorting options and star count display
- Fix table truncation in search command tests

The star count feature helps users identify popular and well-maintained MCP servers
by displaying GitHub stars and supporting sorting by popularity. Network calls are
cached and have timeouts to ensure reliable performance.

slobentanzer committed:
feat: improve relevance scoring and GitHub API integration

- Implement proper relevance scoring algorithm with multi-factor weighting:
  * Exact name match: 10.0 points (highest priority)
  * Partial name match: 8.0 points
  * Exact description match: 5.0 points
  * Partial description match: 3.0 points
  * Exact keyword match: 4.0 points per keyword
  * Partial keyword match: 2.0 points per keyword
  * Star count bonus: stars/100.0 (capped at 3.0) for biomedical repos

- Add GitHub API authentication support to handle rate limiting:
  * Support GITHUB_TOKEN environment variable
  * Support github_token in biotope.yaml config
  * Add proper User-Agent and Accept headers
  * Graceful handling of 403 rate limit responses
  * Fallback to "—" display when rate limited

- Optimize star bonus scaling for biomedical repositories:
  * Change from stars/1000.0 to stars/100.0 for better impact
  * Increase cap from 2.0 to 3.0 points
  * More meaningful differentiation for typical biomedical star counts (10-500)

- Add comprehensive tests for relevance scoring and GitHub token handling
- Update API documentation with relevance algorithm details and GitHub token setup

The relevance scoring now provides intelligent ranking based on content relevance
while giving popular biomedical tools meaningful recognition. GitHub integration
handles rate limiting gracefully with optional authentication for higher limits.

slobentanzer committed:
feat: add bio.tools registry integration to search command

- Add BioToolsRegistry class for bio.tools REST API integration
- Extend search command to support --type biotools for bioinformatics tools
- Implement unified search across all registries when no --type specified
- Add Type column to results table for combined searches
- Include relevance scoring for bio.tools based on name, description, topics, and functions
- Update project configuration to include bio.tools registry settings
- Add comprehensive unit tests for BioToolsRegistry functionality
- Update documentation with bio.tools examples and configuration
- Graceful error handling for individual registry failures
- Smart result counting and action suggestions for different resource types

The search command now provides a unified interface for discovering both MCP servers
and traditional bioinformatics tools, with intelligent relevance ranking across
multiple registries.

slobentanzer committed:
remove docstring for CLI rendering, show help

slobentanzer committed:
display help for incomplete commands, reduce docstrings for CLI display

slobentanzer committed:
feat: enhance bio.tools integration with smart column labeling and composite scoring

- Add smart column labeling: "Stars" for MCP, "Citations" for bio.tools, "Impact" for combined
- Implement composite scoring for bio.tools using citations, validation, homepage, ELIXIR badge, activity
- Change --sort parameter from "stars" to "impact" for unified sorting across registries
- Update search behavior to show combined results by default with registry type indicators
- Improve error handling: graceful degradation instead of hard failures for network issues
- Update test suite to reflect new combined search behavior and CLI help display
- Ensure consistent "citations" key usage in bio.tools data structure (not "stars")
- Add helpful explanations for combined search results showing impact metric differences

The search command now provides a more sophisticated and domain-appropriate experience
with academic impact metrics for bioinformatics tools and community popularity for MCP servers.

slobentanzer committed:
fix: preserve registry ranking in combined search

- Fix default search to respect registry's impact-first ranking
- Remove alphabetical re-sorting that was overriding impact scores
- Ensure high-impact tools (like SCANPY with 4146 citations) appear at top
- Maintain proper ranking for --sort impact and --type biotools searches
- Combined search now preserves registry ranking order instead of overriding it

peymanvahidi commented on biotope/commands/annotate.py:
Sth that I noticed is that we use get_standard_context() function, to get the the Croissant context, but in the create function we're defining the metadata again. I think it's better to use the get_standard_context when creating metadata (line 157)

peymanvahidi commented on biotope/commands/annotate.py:
annotate.py file is already too large. Maybe we can think about how to split it to smaller chunks, so further development would be easier.

peymanvahidi commented on biotope/registry/biotools.py:
```diff
@@ -0,0 +1,327 @@
+            "last_update": last_update
+        }
+
+    def _calculate_composite_score(self, tool: Dict, query: str, all_tools: List[Dict] = None) -> float:
```
We're not using this method anywhere.

peymanvahidi commented on biotope/registry/biotools.py:
```diff
@@ -0,0 +1,327 @@
+            # Normalize scores to 0-1 range
+            max_relevance = max([t["_relevance_score"] for t in results]) if results else 1
+            max_impact = max([t["_impact_score"] for t in results]) if results else 1
+            max_quality = max([t["_quality_score"] for t in results]) if results else 1
```
let's put this part out of the for loop.

peymanvahidi commented on biotope/registry/biotools.py:
```diff
@@ -0,0 +1,327 @@
+        for tool in tools_data:
+            # Ensure identifier is present for ranking
+            tool["identifier"] = tool.get("biotoolsID", tool.get("name", "unknown"))
+            tool["_relevance_score"] = self._calculate_relevance_score(tool, query_lower)
```
I think we should either use _calculate_composite_score here, or add a new sort option to use it.

peymanvahidi commented on biotope/registry/manager.py:
```diff
@@ -0,0 +1,43 @@
+    
+    def fetch_registry(self, url: str, cache_duration: int = 3600) -> list[dict]:
+        """Fetch registry data with caching."""
+        cache_file = self.cache_dir / f"registry_{hash(url)}.json"
```
we should use hashlib.sha256, this hash function produces different outputs if we restart the python session.

peymanvahidi commented on biotope/registry/manager.py:
```diff
@@ -0,0 +1,43 @@
+                    with open(cache_file) as f:
+                        return json.load(f)
+                except (json.JSONDecodeError, IOError):
+                    pass
```
maybe having an error or warning here.

peymanvahidi reviewed (changes requested):
Hey, I reviewed the PR, and these are some minor changes, I think would be beneficial.

slobentanzer commented:
@peymanvahidi many thanks, good points. I agree with the proposed changes. I can implement them, or you do it, but can you please in any case resolve the conflicts with `main` (as these relate to changes you made, and I am not sure I know which branch should be preferred or merged)?

peymanvahidi commented:
@slobentanzer I can handle them, I'll go through them on Monday, I'll resolve the comments I made, as well as the conflicts, and then I'll merge it.

peymanvahidi committed:
refactor: streamline metadata context retrieval in Croissant file creation

- Replace hardcoded metadata context with a call to get_standard_context() for improved maintainability
- Simplify the creation of Croissant metadata files by centralizing context management

peymanvahidi committed:
refactor: optimize score normalization in BioToolsRegistry

- Move max score calculations outside the loop for efficiency

peymanvahidi committed:
feat: add composite scoring option to search command

peymanvahidi committed:
refactor: use sha256 to have the same hashes in different python sessions

peymanvahidi committed:
fix: improve error handling for registry cache loading

peymanvahidi committed:
Merge main into context: integrate batch CSV improvements and keep MCP status feature

peymanvahidi committed:
feat: enhance add and annotate commands to skip biotope's own annotation files

- Updated the add command to skip adding .biotope.csv files during recursive directory processing.
- Removed the duplicated batch annotation command

peymanvahidi committed:
refactor: streamline test setup and metadata handling
